### DISCUSSION TITLE
PocketBase on CloudRun 🎉

### DISCUSSION THREAD

**Discussion by rodydavis** [pocketbase/pocketbase#5054] [Show and tell]
It is now possible to host PocketBase on Google Cloud via CloudRun!

https://github.com/rodydavis/pocketbase-cloudrun

**TLDR**: CloudRun can now mount volumes and this example shows how to easily wire it up.

1. Create a Google Cloud bucket
2. Create service with volume mounted at `/cloud/storage`
3. Set maximum instances to 1
4. Optionally hook into PocketBase health check `/api/health`
5. Fork the example (or use template)
6. Setup cloud build
7. Use the service!

Some awesome things about this deployment is the `pb_public`, `pb_data` and `pb_hooks` are all stored on the bucket. 

Which means you can start locally and rsync the files to the bucket (including the wal files).

Then you use use PocketBase normally and do not have to worry about storage size (or even backups if you setup file revisions).

**Comment by Suryapratap-R** [comment]
very impressive.

have you tried benchmarking it?

**Reply by rodydavis** [reply]
Not yet, but works great in my testing, and randomly checking in on it

**Reply by stewart-codes** [reply]
I’m a fan of cloud run so I would be really happy if this setup works. I am curious how it will perform on a larger database. Based on this snippet from the FUSE docs, it will have to store the whole db file to the bucket for every write. Sounds like that won’t scale for a write heavy application.

> Workloads that do file patching (or overwrites in place): Cloud Storage FUSE can only write whole objects at a time to Cloud Storage and does not provide a mechanism for patching. If you try to patch a file, Cloud Storage FUSE will reupload the entire file. The only exception to this behavior is that you can append content to the end of a file that's 2 MB and larger, where Cloud Storage FUSE will only reupload the appended content.

https://cloud.google.com/storage/docs/gcs-fuse

**Reply by stewart-codes** [reply]
Reading further in the linked doc is says to not use this as a database file system 😅

> Performance: Cloud Storage FUSE has much higher latency than a local file system, and as such, shouldn't be used as the backend for storing a database.

**Reply by rodydavis** [reply]
But it WAL mode it's only appending right?

**Reply by stewart-codes** [reply]
From a little reading I just did [1], the answer is no. The WAL file is indeed append only. But in WAL mode, there's 2 sqlite files, the main database file and the WAL file. At checkpoints, the WAL file changes are written to the main db which is not append only. Theoretically one could take advantage of the WAL buffer and the 2 MB FUSE append only optimization. But there's not way to get around writing back to the main db file. Keep in mind I'm just theorizing here. Maybe it actually does work like you said. Some performance tests would really need to be run to understand how a production workload performs.

[1] https://sqldocs.org/sqlite/sqlite-write-ahead-logging/

**Reply by rodydavis** [reply]
It definitely works! Just need to test how many requests. For mostly read only and single admin updating content like a blog it should be fine

**Reply by stewart-codes** [reply]
That’s a perfect use case for this setup. I was considering a typical read + write use case.

**Comment by pchasco** [comment]
Did you mean to say "It is NOW possible to host PocketBase on Google Cloud via CloudRun?"

**Reply by rodydavis** [reply]
Yes! Edited 😅

**Comment by tigawanna** [comment]
this one is super nice , i just had 2 questions
does it haveto be mounted at `/cloud/storage` and can one mount it in a sub directory or is it limited to 1 bucket per pocketbase service

**Reply by rodydavis** [reply]
You can mount as many buckets as you want! And you define the path to be mounted in, I just thought that was a good path name!

**Reply by tigawanna** [reply]
oh , what about 1 bucket 2 projects each with it's own sub-directory

**Reply by rodydavis** [reply]
Yep totally possible!

**Comment by zeshhaan** [comment]
wow! this is such a good news! Is it possible to use for production apps?

**Reply by rodydavis** [reply]
Depends on how write heavy the application is

**Reply by zeshhaan** [reply]
I have successfully deployed. but when I try to access the url, it gives me 404 but health check is giving 200. Maybe I'm missing the right paths

<img width="2032" alt="image" src="https://github.com/user-attachments/assets/52894221-b9eb-470f-9dcf-a788f542039c">
<img width="473" alt="image" src="https://github.com/user-attachments/assets/ccfe79a5-a8ac-4610-8ef1-03def54b9d8f">

**Reply by rodydavis** [reply]
It's working!

Just do /_ at then end of the url

**Reply by zeshhaan** [reply]
oh yeah! i forgot about that. woohoo!

**Reply by rodydavis** [reply]
With PocketBase those routes will go to the pb_public static directory by default. So you can add a pb_hook with a redirect or add an index.html

**Reply by zeshhaan** [reply]
thanks so much!

**Comment by benallfree** [comment]
@rodydavis This is interesting, have you gained any further insight to using FUSE-backed persistent storage? I see you can route all requests to a single machine, that's good. Some questions:

1. What happens when that machine exits? Will it wait until all data is flushed back to object storage? 
2. And while it is shutting down, is it still available for new requests?

> or even backups if you setup file revisions

I'm not sure this is correct, as [FUSE docs](https://cloud.google.com/storage/docs/gcs-fuse#expandable-1) state `Attempting to use Cloud Storage FUSE with buckets that have object versioning enabled can produce unpredictable behavior.`
